### PR TITLE
Fix reptile_step implementation

### LINKED ISSUE #14980: reptile_step

### PR DESCRIPTION
**Pull request by abbasaliii**
This pull request fixes the implementation of the reptile_step function in the meta.py file. The function was refactored according to the Deep Dive guidelines to improve its efficiency and functionality.

Changes Made:

Refactored the reptile_step function to enhance its performance and readability.
Addressed issues with variable assignments and function arguments.
Optimized the inner loop optimization process by adjusting gradient steps and learning rate.
Added support for batching along the time dimension for improved meta steps.
These changes aim to ensure that the reptile_step function operates correctly and efficiently within the larger codebase. The modifications have been tested and verified to provide the desired behavior.

Fixes #14980

Please review these changes and consider merging them into the main codebase. Thank you for your attention and feedback.

### CODE REVIEW

File: ivy/functional/ivy/meta.py [hunk 1]
```diff
@@ -533,58 +533,95 @@ def fomaml_step(
...
     variables: ivy.Container,
     inner_grad_steps: int,
     inner_learning_rate: float,
     /,
     *,
-    inner_optimization_step: Callable = gradient_descent_update,
+    inner_optimization_step: Callable[
+        [Callable[[ivy.Array, ivy.Container], ivy.Array], ivy.Container, float],
+        ivy.Container,
+    ] = gradient_descent_update,
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/meta.py:545]: I appreciate the detail you've added to these type hints, but I think we should leave them as just 'Callable' so they are less verbose and follow the same pattern as other functions.

File: ivy/functional/ivy/meta.py [hunk 2]
```diff
@@ -533,58 +533,95 @@ def fomaml_step(
...
     ----------
-    batch
-        The input batch
-    cost_fn
-        callable for the cost function, receivng the task-specific sub-batch and
-        variables
-    variables
-        Variables to be optimized
-    inner_grad_steps
+    batch : ivy.Container
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/meta.py:560]: There's no need to include type hints in the docstring

File: ivy/functional/ivy/meta.py [hunk 3]
```diff
@@ -533,58 +533,95 @@ def fomaml_step(
...
-        Whether to stop the gradients of the cost. Default is ``True``.
+        the batch by default.
+    stop_gradients : bool, optional
+        Whether to stop the gradients of the cost. Default is `True`.
 
     Returns
     -------
-    ret
-        The cost and the gradients with respect to the outer loop variables.
+    Tuple[ivy.Array, ivy.Container, Any]
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/meta.py:593]: Similarly, this should be left as 'ret'

File: ivy/functional/ivy/meta.py [hunk 4]
```diff
@@ -533,58 +533,95 @@ def fomaml_step(
...
+    The `inner_optimization_step` parameter specifies the function used for the inner
+    loop optimization. It takes the cost function, variables, and learning rate as
+    arguments and returns the updated variables. The default function
+    `gradient_descent_update` performs gradient descent using the provided learning
+    rate.
+
+    The `return_inner_v` parameter controls which variables are returned from the inner
+    loop optimization. If set to `'first'`, the variables for the first task inner loop
+    will also be returned. If set to `'all'`, variables for all tasks will be returned.
+    By default, only the cost and gradients are returned.
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/meta.py:620]: Please can you add some docstring examples here? You can find more info on them here: https://unify.ai/docs/ivy/overview/deep_dive/docstring_examples.html. The docstring examples from this stale reptile_step PR would be a great start point https://github.com/unifyai/ivy/pull/15120

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@abbasaliii Thanks for the PR! I've added some comments for things we need to change, please could you take a look?

**Comment by abbasaliii** [comment]
@Sam-Armstrong 

Thank you for your feedback and comments on the changes made to the reptile_step function. I have addressed the issues you mentioned and made the following updates:

Simplified the type hints to use just 'Callable' instead of the verbose versions.
Removed the type hints from the docstring to align with the standard practice.
Updated the return value to 'ret' instead of explicitly stating the tuple type.
Please review the changes and let me know if there is anything else that needs to be addressed. Your input is greatly appreciated.

File: ivy/functional/ivy/meta.py [hunk 5]
```diff
@@ -530,25 +530,29 @@ def fomaml_step(
...
-    ] = gradient_descent_update,
+    inner_optimization_step: Callable = gradient_descent_update,
     batched: bool = True,
     return_inner_v: Union[str, bool] = False,
     num_tasks: Optional[int] = None,
     stop_gradients: bool = True,
-) -> Tuple[ivy.Array, ivy.Container, Any]:
+):
     """
+<<<<<<< HEAD
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/meta.py:555]: I think something has gone wrong with your merge conflict resolution - this HEAD should have been removed

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@abbasaliii I think something went wrong with your merge, please can you have a look?

**Comment by abbasaliii** [comment]
@Sam-Armstrong , I have made the requested changes to the code. Please take a look and let me know if there's anything else I can assist you with.

**Comment by Sam-Armstrong** [comment]
@abbasaliii Please can you add some docstring examples? You can find more info on them here: https://unify.ai/docs/ivy/overview/deep_dive/docstring_examples.html. The docstring examples from this stale reptile_step PR would be a great start point https://github.com/unifyai/ivy/pull/15120

**Comment by abbasaliii** [comment]
@Sam-Armstrong, I have added docstring examples

**Comment by Sam-Armstrong** [comment]
@abbasaliii apologies, I've been really busy, will review later today 😁
